move same_unit() in unit class

newfrenchy83 opened the pull request.

stevecotton commented on src/units/unit.hpp:
```diff
@@ -1897,6 +1897,14 @@ class unit : public std::enable_shared_from_this<unit>
 	 */
 	bool ability_matches_filter(const config & cfg, const std::string& tag_name, const config & filter) const;
+	/**
+	 * Verify what unit checked is not itself.
```
Please copy the comment from abilities.cpp, it's clearer and includes the "same location" part.

Vultraz commented:
Instead of moving this function to the unit class, I'd suggest creating a function that returns a filtered view over the units. That would remove the loop check that is repeated multiple times. I can handle it if you'd like.

newfrenchy83 commented:
> Instead of moving this function to the unit class, I'd suggest creating a function that returns a filtered view over the units. That would remove the loop check that is repeated multiple times. I can handle it if you'd like.

i'm not sure to understand

newfrenchy83 commented:
> Instead of moving this function to the unit class, I'd suggest creating a function that returns a filtered view over the units. That would remove the loop check that is repeated multiple times. I can handle it if you'd like.

https://github.com/wesnoth/wesnoth/pull/10430 Does this PR match your suggestion?

stevecotton commented:
> I can handle it if you'd like.

@newfrenchy83: Vultraz is offering to write code that works in the way that he wants it to. Please wait for him to do so.

@Vultraz it sounds a good offer to me.

newfrenchy83 commented:
> > I can handle it if you'd like.
> 
> @newfrenchy83: Vultraz is offering to write code that works in the way that he wants it to. Please wait for him to do so.
> 
> @Vultraz it sounds a good offer to me.

ok, then.

newfrenchy83 commented:
@soliton- @Vultraz I reviewed udisplay.cpp, when the movement is in progress, the system switches between the unit and its fake when passing from one hex to another, relocates the unit then redisplays the unit and hides the fake, and so on until the arrival, this is why fake and original are always next to each other, and as fake_unit_manager is also involved in the display and game_display with the cursor, it is more dangerous to modify the code of fake_unit than to solve the problem at the level of abilities and filters since it is the abilities that need the check of the identifier to ensure that the unit reviewed is not the owner while looking for an adjacent/distant unit.

And sorry for my outburst, but your insistence on looking at the fake_unit code rather than looking at this PR again is starting to annoy me a little.

Vultraz commented:
I also went and looked at the fake unit code too. It seems odd that the unit_mover clones the whole unit and hides the original, but I can see now why the simple `&u == this` check didn't work. Even though there isn't a new unit in the unit map, the temp unit (that is, `this`, which is the cloned unit object owned by the unit_mover) is its own unique object.

That being said, the current state of this PR is an improvement over the original. I would suggest comparing against `underying_id` rather than `id`, since as far as I can tell that's a safer "is the same unit" check than just ID (as you say, just ID might not be a strict enough check).

newfrenchy83 commented:
@soliton- Okay, I'll try to explain again.

When calling an ability function, we first check if the unit has an ability of the specified type that applies to itself, then we eventually move on to checking all the units on the map to see if they have an ability with [affect_adjacent], except that in this context, the unit (this) doesn't need to be rechecked, and when a unit on the map turns out to be 'this', we move directly to the next one (before moving on to the distant 'radius' attribute, we checked the adjacent hexes and verified that a unit was there).

The problem comes from the comparison method I had chosen to be the most rigorous: if u == *this , where the entire unit was compared.

The problem, as @Vultraz understood, is that the fake_unit class is a separate object, and in the case of movement, the generated fake_unit and the original cease to be identical because their location ceases to be identical, which causes the check to fail.

For example, if we take a unit with an ambush ability that only applies to adjacent units, when we move this unit onto favorable terrain, the unit will be shown as concealed during the movement, even though it shouldn't be, because another 'unit' is there, the fake_unit, and both are affected by their neighbor.

To remedy this, I created the same_unit() function to compare only locations and id, and now I simply try to use the underlying_id as the sole criterion to achieve a reliable comparison in all circumstances.

As for doing without fake_unit, I've already tried it, it just makes the game crash.

stevecotton commented:
> When calling an ability function, we first check if the unit has an ability of the specified type that applies to itself, then we eventually move on to checking all the units on the map to see if they have an ability with [affect_adjacent], except that in this context, the unit (this) doesn't need to be rechecked, and when a unit on the map turns out to be 'this', we move directly to the next one (before moving on to the distant 'radius' attribute, we checked the adjacent hexes and verified that a unit was there).

There are different types of documentation, and the documentation should tell other developers things that aren't obvious from the code. Consider a function that "Returns true if unit1 and unit2 are the same unit, including when either is a fake_unit generated from the other one. Used to prevent fake_units being shown as affected by their own [affects_adjacent] abilities.". That's still not well-worded (I'll post a better suggestion if I think of one later), but it hints to other developers about why the function exists.

soliton- commented:
> As for doing without fake_unit, I've already tried it, it just makes the game crash.

Everything written above this is what was clear by now. This is the only sentence of interest but unfortunately it explains very little. What did you try in detail? What issues did you face?

Also no response to what other code will need to be adjusted if we keep with the fake unit logic leaking into unrelated code. Or if no other code needs adjustment, based on what can that be verified?

stevecotton commented on src/units/unit.hpp:
```diff
@@ -1898,6 +1898,33 @@ class unit : public std::enable_shared_from_this<unit>
 	bool ability_matches_filter(const config & cfg, const std::string& tag_name, const config & filter) const;
 
+	/**
+	 * Return true, if unit calling function and unit designated by @ self_underlying_id are different units including when either is a fake_unit generated from the other one, if unit not incapacitated and own a ability with [affect_adjacent] subtag
```
This doesn't feel right to me, it's putting together three different concepts.

The id check is commutative - if A has a different id to B then B has a different id to A. The "when either ... from the other one" text suggests that it doesn't matter which unit is A and which is B, which is true for the id part, but false for the incapacitated() and has_ability_distant() checks.

Having a function that combines `!incapacitated() && has_ability_distant()` doesn't feel right to me either - when it's checking `incapacitated()` it's handling one way in which the ability might be inactive, but not others. It makes sense that the three checks would be on the same line in the caller, but I feel they aren't similar enough to put them into a separate function.

soliton- commented:
> I tried modifying the udisplay.cpp code to directly modify the 'unit' variable instead of using a fake_unit, but as soon as I tried to move a unit with that, the game simply crashed.

Did you do any kind of debugging? What exactly did your modifications look like?

> Aside from the ability-related issues, there are no other issues I'm aware of with using fake_unit, which is why I prefer to fix the map unit checking in the ability functions instead.

Have you looked for other issues? Ideally not just be trying random stuff when playing wesnoth but by reading code and thinking about what's possible.

The abilities and filter code is used in lots of other contexts. I do not understand how you're so eager to change that code just because the animation code does something special.

I'd much rather understand the need for fake units first (and document it). If there even is a need still what exactly is it? I can guess that maybe some unit state needs to be modified for the animation code that somehow cannot be reverted easily maybe?

I think simply trying to not use fake units at all and see what happens is an ok first step. The conclusion cannot possibly be "the game simply crashed" though. I'm a bit lost for words here and probably should not look for some...

stevecotton commented:
I also want to know why we can't remove the fake unit code from the animation.

gfgtdf commented:
>I'd much rather understand the need for fake units first 

One of the main reason i'd think of is that we currently cannot just change the original units position on the map, for example in case that a unit moves over another unit during movement the location might already be blocked.

Adding direct support for showing movement on a different location without fake units would certainly also make certain things easier tho (i'm thinking in particular of allowing a good implementation of #9487), i think it might be rather huge taks to implement tho. (also not really related to this pr)




I agree tho that it should be documented better why the code does what it currently does, at the very least in this commits commit message.

I also think that this commit (which needs to change ~10 for loops of extremely similar shape to do a change) shows how much code duplication is already in this messy file, it already was kinda hard to maintain 5 years ago but imo it got worse with all the recent additions to abilities, which is also why i usually don't review related prs anymore.

stevecotton commented on src/units/abilities.cpp:
```diff
@@ -199,11 +199,21 @@ int find_direction(const map_location& loc, const map_location& from_loc, std::s
 
 /**
- * This function return true if locations checked are the same, or if units have same id.
+ * These functions return true if @a u not incapacitated, if @a u underlying_id() not equals to *this underlying_id() and other critera.
```
I refer to my [previous comment](https://github.com/wesnoth/wesnoth/pull/10426#discussion_r2257689727), that it doesn't feel right putting these different concepts together in a function.

Neither the documentation nor the name hint at why a unit would be "valid" when maintaining this file, so reading it without the benefit of seeing that it's part of a PR about affect_adjacent abilities.

stevecotton commented on src/units/abilities.cpp:
```diff
@@ -199,11 +199,19 @@ int find_direction(const map_location& loc, const map_location& from_loc, std::s
+ */
+bool unit_has_ability_distant(const std::optional<int>& has_ability_distant, const std::size_t& distance)
+{
+	return(has_ability_distant && has_ability_distant <= distance);
```
Should this be `distance <= has_ability_distance`?

I'm wondering why `unit::has_ability_distant()` returns `std::optional<std::size_t>` instead of a plain `std::size_t` with zero meaning "no abilities that affect units one or more hexes away".
```
	/**
	 * If this unit has abilities with [affect_adjacent] subtags, returns the radius of the one with the furthest reach.
	 *
	 * If the unit has no such abilities, returns zero.
	 */
	std::size_t max_ability_radius() const
```

stevecotton replied:
Note: zero / non-zero will automatically convert to bool as needed.

stevecotton commented:
My gut feeling is that the simplification commit doesn't improve the readability of the code. However, a review should also have a constructive suggestion about what to do instead, and at the moment I'm lacking an answer to the question "what would simplify this file?".

Certain kinds of documentation would help, but I'm failing to communicate what kind.

@newfrenchy83, if it would be easier for you to write in French then I'll happily read it.

stevecotton commented on src/units/abilities.cpp:
```diff
@@ -228,12 +220,9 @@ bool unit::get_ability_bool(const std::string& tag_name, const map_location& loc
 	// different from the central unit, that the ability is of the right type, detailed verification of each ability),
 	// if so return true.
 	for(const unit& u : units) {
-		if(!u.has_ability_distant() || u.incapacitated() || same_unit(u, *this) || !u.affect_distant(tag_name)) {
```
Some of these checks will be faster than others, some will return false more often than others, and although they're all going to be fast, I assume this code path is going to be heavily used during the AI turn. Assuming most units won't have `[affects_adjacent]` abilities, I feel it's good to check `u.has_ability_distant()` before even calculating `distance_between`.

newfrenchy83 commented:
> My gut feeling is that the simplification commit doesn't improve the readability of the code. However, a review should also have a constructive suggestion about what to do instead, and at the moment I'm lacking an answer to the question "what would simplify this file?".
> 
> Certain kinds of documentation would help, but I'm failing to communicate what kind.
> 
> @newfrenchy83, if it would be easier for you to write in French then I'll happily read it.

I remade this PR in three commits. In the first, I simply replaced the comparison between the unit checked for [affect_adjacent] and *this in same_unit() with a comparison using underlying_id(). In the second, I renamed the unit variables involved to max_ability_radius and changed the declaration. So far, everything is clear.

For the last commit, the functions managing abilities used as weapons returning bool values and those for filtering specials tend to reuse the same code for self_ and other_. So, I used two impl functions to simplify the writing, and thus the number of for(const unit u : units) loops is reduced from 11 to 7, which will be a little less redundant to read.

stevecotton commented:
@soliton- I've spent a while looking at the fake unit handling in the animation code, which has made me accept that we need to handle fake units in the abilities code, and @Vultraz has already approved the superset of these changes that's in the new PR. My feeling is that we merge this one when the CI finishes, and then discuss how to deal with the already-existing code readability problems in the new PR.

My notes are the commit messages in https://github.com/wesnoth/wesnoth/compare/master...stevecotton:wesnoth:move_without_fake_unit

newfrenchy83 commented:
> @soliton- I've spent a while looking at the fake unit handling in the animation code, which has made me accept that we need to handle fake units in the abilities code, and @Vultraz has already approved the superset of these changes that's in the new PR. My feeling is that we merge this one when the CI finishes, and then discuss how to deal with the already-existing code readability problems in the new PR.
> 
> My notes are the commit messages in [master...stevecotton:wesnoth:move_without_fake_unit](https://github.com/wesnoth/wesnoth/compare/master...stevecotton:wesnoth:move_without_fake_unit)

CI finished
